domain/network/state/movesubnets.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/transform"
+
+	"github.com/juju/juju/core/network"
+	domainnetwork "github.com/juju/juju/domain/network"
+	"github.com/juju/juju/internal/errors"
+)
+
+// MoveSubnetsToSpace moves the specified subnets to a given network space,
+// enforcing space constraints if applicable.
+func (st *State) MoveSubnetsToSpace(
+	ctx context.Context,
+	subnetUUIDs []string,
+	spaceName string,
+	force bool,
+) ([]domainnetwork.MovedSubnets, error) {
+	if len(subnetUUIDs) == 0 {
+		return nil, nil
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var positiveFailures []positiveSpaceConstraintFailure
+	var negativeFailures []negativeSpaceConstraintFailure
+	var movedSubnets []domainnetwork.MovedSubnets
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := st.checkSpaceName(ctx, tx, spaceName)
+		if err != nil {
+			return errors.Capture(err)
+		}
+		subnetToMove := uuids(subnetUUIDs)
+		positiveFailures, err = st.checkMovedFromMachine(ctx, tx, subnetToMove, spaceName)
+		if err != nil {
+			return errors.Capture(err)
+		}
+		negativeFailures, err = st.checkMovedToMachine(ctx, tx, subnetToMove, spaceName)
+		if err != nil {
+			return errors.Capture(err)
+		}
+		if err := st.handleFailures(ctx, positiveFailures, negativeFailures, spaceName, force); err != nil {
+			return err
+		}
+		movedSubnets, err = st.moveSubnets(ctx, tx, subnetToMove, spaceName)
+		return errors.Capture(err)
+	})
+
+	return movedSubnets, errors.Capture(err)
+}
+
+// checkMovedFromMachine verifies if subnets moved to a given space violate
+// positive space constraints for any machines.
+// It queries the database to identify machines whose space constraints
+// conflict with the updated subnet-to-space mapping.
+// Returns a list of positiveSpaceConstraintFailure instances or an error
+// if the process fails.
+func (st *State) checkMovedFromMachine(
+	ctx context.Context,
+	tx *sqlair.TX,
+	movedSubnets uuids,
```

Comment:
I don't think these methods are really machine centric.

The names should be `validateSubnetsJoiningSpace` and `validateSubnetsLeavingSpaces`